### DISCUSSION TITLE
Table Update does not update file

### DISCUSSION THREAD

**Discussion by aebaranoff** [nasa/cFS#708] [Q&A] [answered]
I have a table I have updated. But it does not update (save back to disk).

After I update if I call CFE_TBL_Update it says CFE_TBL_INFO_NO_UPDATE_PENDING.  I definitely updated the table but for some reason it does not pick up the update. 

The table is stored in the location returned by CFE_TBL_GetAddress. The table is initially loaded successfully but will not save.

I tried adding in CFE_TBL_DumpToBuffer and CFE_TBL_Manage but I am either missing a step or doing the stepping in the wrong order.

This is the draco-rc4 version of cFE,

**Comment by BBinLP** [comment]
Did you remember to "Activate" the table?

-David

**Comment by BBinLP** [comment]
To be more clear,

One first "Loads" a table.  This moves the data from a file into a
temporary buffer.
One can then "Verify" the table.  This allows the application to look at
the contents of the temporary buffer to make sure it is logically sound for
its purpose.
Then you "Activate" the table.  At this point, Table Services will notify
the application the next time it calls "Manage" and the application will
update the contents of the table.

-David

**Comment by aebaranoff** [comment]
How do I activate? I am new to cFE/CFS.

Thanks,

**Comment by BBinLP** [comment]
There should be an Activate command in your command database for your ground system to do it.

**Comment by aebaranoff** [comment]
I want to activate it on the payload.

**Comment by BBinLP** [comment]
Ah, I have been misunderstanding what you were trying to do.

Is the application modifying the table the same one that uses the table?  Or are you trying to modify the table of Application "A" with Application "B"?

**Comment by aebaranoff** [comment]
Based on HW interaction we want to save a table updates. Also, if the ground commands it we would change the table and then that would update the flash file copy of the table.

**Comment by BBinLP** [comment]
Sounds like your paradigm might be a little different from what Table services was originally designed to do.  However, it can still be done but you may need to send commands over the software bus to the CFE TBL application to make it work correctly.

Let me see if I understand what you are trying to do:

An application is registering a table with table services with a call to CFE_TBL_Register().
It loads a default image into that table from a block of memory or from a file using CFE_TBL_Load().
Then, during routine execution, the application calls CFE_TBL_GetAddress() to get an address of the table.
It accesses the contents of the table and may even change the contents of the table and then before it waits again for its next cycle of execution, it releases the address using CFE_TBL_ReleaseAddress().  Finally, your application calls CFE_TBL_Manage() to handle table load, verify, and dump requests from the CFE TBL application.

If this is the general usage of the table, you should also be sure to call CFE_TBL_Modified() whenever you have changed the contents.  This will result in a flag in telemetry letting the ground operator know that the table image is no longer the same as the original source.  It also makes it clear to table services that the table has changed and if it is a Critical table, it will update the copy kept in the Critical Data Store.

Now, I think you are saying that you want to save the table updates to some non-volatile location (e.g. file storage).  If you want to save it to a file on your file system, I would form a command in your application to the CFE TBL application that commands CFE TBL to "Dump" the table to a filename of your choice and send that as a message on the software bus.  CFE TBL services will then communicate back to your application (via the CFE_TBL_Manage() API) and dump the contents of the table into the file you specified.

**Reply by aebaranoff** [reply]
Here is the with proprietary stuff removed...

/* this is how I initialized stuff -  I have verified we initially load the table from the local file system */
int TableInit()
{
    //Register tables with table services, then load table data
    int32 status = CFE_SUCCESS;
    status = CFE_TBL_Register(&My_Data.MyTableHandle, "My_config", sizeof(MyConfigTable_t), CFE_TBL_OPT_DEFAULT, &ValidateTable);
    if(status == CFE_SUCCESS)
    {
        status = CFE_TBL_Load(My_Data.MyTableHandle, CFE_TBL_SRC_FILE, My_CONFIG_TABLE_DEFAULT_FILE);
        CFE_TBL_Manage(My_Data.MyTableHandle);
        
        status = CFE_TBL_GetAddress((void*)&My_Data.MyTable, My_Data.MyTableHandle);

        CFE_Status_t res;
        res = CFE_TBL_Validate(My_Data.MyTableHandle);
        printf("CFE_TBL_Validate %d\n", res);


        if(status == CFE_TBL_ERR_NEVER_LOADED)
        {
            CFE_EVS_SendEvent(My_CONFIG_TABLE_INVALID_EID, CFE_EVS_EventType_ERROR, "My Table Never Loaded");
            status = CFE_SUCCESS;
         }
    }
    return status;
}


/* This is the validation logic - everything passes unless I force an error */
int ValidateTable(void* TableData)
{

	...
    return CFE_SUCCESS;
}

// This is where I am trying to save the data it has changed before getting here 
int TableUpdate(t_MyCfgTblCore* TableData)
{
    CFE_Status_t res;
    CFE_StatusString_t teststr;

    My_Data.MyTable->Config[TableData->index] = *TableData;
    My_Data.MyTable->Config[TableData->index].test = My_Data.MyTable->Config[TableData->index].test-1;

    res = CFE_TBL_Validate(My_Data.MyTableHandle);
    printf("CFE_TBL_Validate %d %s\n", res, CFE_ES_StatusToString(res,&teststr));    

    res = CFE_TBL_GetStatus(My_Data.MyTableHandle);
    printf("CFE_TBL_GetStatus %d %s\n", res, CFE_ES_StatusToString(res,&teststr));    
    res = CFE_TBL_Update(My_Data.MyTableHandle);
    printf("TableUpdate %d %s\n", res, CFE_ES_StatusToString(res,&teststr));
    res = CFE_TBL_DumpToBuffer(My_Data.MyTableHandle);
    printf("CFE_TBL_DumpToBuffer %d %s\n", res, CFE_ES_StatusToString(res,&teststr));    
    res = CFE_TBL_Manage(My_Data.MyTableHandle);
    printf("CFE_TBL_Manage %d %s\n", res, CFE_ES_StatusToString(res,&teststr));    
    return CFE_SUCCESS;
}

**Reply by BBinLP** [reply, marked as answer]
In the initialization section, the code following CFE_TBL_Manage shouldn't be needed.  CFE_TBL_Manage will do any verification if there is any that needs to be done.  The status returned by CFE_TBL_Load will let you know whether the file was successfully loaded into the table or not.

In the TableUpdate function, the first assignment statement may be copying the contents of the table into Config[TableData->index]. You are now manipulating the contents of a copy rather than the actual active table managed by Table Services.  I would suggest the code looking something like the following:
`
TableData->test = TableData->test - 1;`

Then notify Table Services that the contents have been modified:

`CFE_TBL_Modified(My_Data.MyTableHandle);`

To send a command to table services to dump the table to a file, you need to create an instance of the following command structure in your application data structure:

```
typedef struct
{
    CFE_MSG_CommandHeader_t  CommandHeader;
    uint16 ActiveTableFlag;
    char    TableName[CFE_MISSION_TBL_MAX_FULL_NAME_LEN];
    char    DumpFilename[CFE_MISSION_MAX_PATH_LEN];
} t_CfeTblDumpCmd
```
You need to initialize the message header during application initialization along with the command secondary header:

```
CFE_MSG_Init(&My_Data.CfeTblDumpCmd. CFE_SB_ValueToMsgId(CFE_TBL_CMD_MID), sizeof(t_CfeTblDumpCmd));
CFE_MSG_SetFcnCode(&My_Data.CfeTblDumpCmd.CommandHeader, CFE_TBL_DUMP_CC);
```
Now, back in your TableUpdate function, you want to populate the rest of the command with the appropriate values for your table:

```
My_Data.CfeTblDumpCmd.ActiveTableFlag = CFE_TBL_BufferSelect_ACTIVE;
strncpy(&My_Data.CfeTblDumpCmd.TableName, "MyApp.TableName", sizeof(My_Data.CfeTblDumpCmd.TableName));
strncpy(&My_Data.CfeTblDumpCmd.DumpFilename, "/path/to/file/filename.tbl", sizeof(My_Data.CfeTblDumpCmd.DumpFilename));
```

And finally, send the command onto the software bus to CFE TBL Services:

`CFE_SB_TransmitMsg(CFE_MSG_PTR(&My_Data.CfeTblDumpCmd.CommandHeader), false);`

When Table Services receives this command, it will, in turn, notify the application when it next calls CFE_TBL_Manage, that the table needs to be dumped and all the steps necessary to perform that dump to the specified file will be taken care of at that time.

**Reply by aebaranoff** [reply]
Thank you for your help.
